### ISSUE TITLE
A memory leak in Windows

### ISSUE THREAD

**Issue by oherrala** [tokio-rs/mio#919]
The following code can be used to trigger a memory leak in Windows. I first found this with `trust-dns-client` (see https://github.com/bluejekyll/trust-dns/issues/692), but the issue can be isolated into `mio` only.

```toml
[package]
name = "mio-mao"
version = "0.1.0"
authors = ["Ossi Herrala <oherrala@iki.fi>"]
edition = "2018"

[dependencies]
env_logger = "*"
mio = "0.6.16"
```

```rust
use std::io;
use std::net::{SocketAddr, Ipv4Addr};
use std::thread::sleep_ms;

use env_logger;
use mio::{Poll, Token, PollOpt, Ready};
use mio::net::UdpSocket;

fn main() -> io::Result<()> {
    env_logger::init();

    let target = SocketAddr::new(Ipv4Addr::new(192,168,1,1).into(), 53);
    let buf = &[0x00; 1400];
    let socket = UdpSocket::bind(&SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0))?;

    let poll = Poll::new()?;
    poll.register(&socket, Token(0), Ready::readable(), PollOpt::edge())?;

    socket.send_to(buf, &target)?;
    sleep_ms(1_000);

    Ok(())
}
```

Trace level debug output of a run:

```
[2019-02-25T12:21:48Z TRACE mio::poll] registering with poller
[2019-02-25T12:21:48Z TRACE mio::sys::windows::selector] register Token(0) Readable
[2019-02-25T12:21:48Z TRACE mio::sys::windows::selector] set readiness to (empty)
[2019-02-25T12:21:48Z TRACE mio::sys::windows::udp] scheduling a read
[2019-02-25T12:21:48Z TRACE mio::sys::windows::selector] set readiness to (empty)
[2019-02-25T12:21:48Z TRACE mio::sys::windows::udp] scheduling a send
```

And Dr. Memory results of the leak:

```
Dr. Memory version 1.11.0 build 2 built on Aug 29 2016 02:41:18
Dr. Memory results for pid 5244: "mio-mao.exe"
Application cmdline: "C:\Users\IEUser\AppData\Local\Temp\cargo\debug\mio-mao.exe"
Recorded 115 suppression(s) from default C:\Program Files (x86)\Dr. Memory\bin64\suppress-default.txt

Error #1: LEAK 352 direct bytes 0x000001d9588e2c90-0x000001d9588e2df0 + 137680 indirect bytes
# 0 replace_RtlAllocateHeap                                            [d:\drmemory_package\common\alloc_replace.c:3770]
# 1 ntdll.dll!RtlPcToFileHeader                                       +0x1e     (0x00007fff43500f6f <ntdll.dll+0x10f6f>)
# 2 ntdll.dll!LdrAddRefDll                                            +0x4b     (0x00007fff4354913c <ntdll.dll+0x5913c>)
# 3 core::cell::UnsafeCell<>::new<>                                    [/rustc/9fda7c2237db910e41d6a712e9a2139b352e558b\src\libcore\cell.rs:1471]
# 4 std::alloc::__default_lib_allocator::__rdl_alloc                   [/rustc/9fda7c2237db910e41d6a712e9a2139b352e558b\/src\libstd\alloc.rs:243]
# 5 alloc::alloc::alloc                                                [/rustc/9fda7c2237db910e41d6a712e9a2139b352e558b\src\liballoc\alloc.rs:82]
# 6 alloc::alloc::exchange_malloc                                      [/rustc/9fda7c2237db910e41d6a712e9a2139b352e558b\src\liballoc\alloc.rs:192]
# 7 core::sync::atomic::AtomicUsize::new                               [/rustc/9fda7c2237db910e41d6a712e9a2139b352e558b\src\libcore\sync\atomic.rs:1158]
# 8 mio::sys::windows::from_raw_arc::FromRawArc<>::new<>               [C:\Users\IEUser\.cargo\registry\src\github.com-1ecc6299db9ec823\mio-0.6.16\src\sys\windows\from_raw_arc.rs:41]
# 9 mio::sys::windows::udp::UdpSocket::new                             [C:\Users\IEUser\.cargo\registry\src\github.com-1ecc6299db9ec823\mio-0.6.16\src\sys\windows\udp.rs:61]
#10 mio::net::udp::UdpSocket::from_socket                              [C:\Users\IEUser\.cargo\registry\src\github.com-1ecc6299db9ec823\mio-0.6.16\src\net\udp.rs:139]
#11 log::set_max_level                                                 [C:\Users\IEUser\.cargo\registry\src\github.com-1ecc6299db9ec823\log-0.4.6\src\lib.rs:1059]

===========================================================================
FINAL SUMMARY:

DUPLICATE ERROR COUNTS:

SUPPRESSIONS USED:

ERRORS FOUND:
      0 unique,     0 total unaddressable access(es)
      0 unique,     0 total invalid heap argument(s)
      0 unique,     0 total GDI usage error(s)
      0 unique,     0 total handle leak(s)
      0 unique,     0 total warning(s)
      1 unique,     1 total, 138032 byte(s) of leak(s)
      0 unique,     0 total,      0 byte(s) of possible leak(s)
ERRORS IGNORED:
      4 potential error(s) (suspected false positives)
         (details: C:\Users\IEUser\AppData\Roaming\Dr. Memory\DrMemory-mio-mao.exe.5244.000\potential_errors.txt)
      1 potential leak(s) (suspected false positives)
         (details: C:\Users\IEUser\AppData\Roaming\Dr. Memory\DrMemory-mio-mao.exe.5244.000\potential_errors.txt)
     22 unique,    22 total,   3609 byte(s) of still-reachable allocation(s)
         (re-run with "-show_reachable" for details)
Details: C:\Users\IEUser\AppData\Roaming\Dr. Memory\DrMemory-mio-mao.exe.5244.000\results.txt
```

**Comment by oherrala** [comment]
My investigation so far:

Windows side of UdpSocket uses overlapped memory to communicate with IOCP. This seem to include `mem::forget()` for the `UdpSocket`'s `FromRawArc<Imp>`.

Tcp side documentation:

https://github.com/carllerche/mio/blob/f2c110363c2acc527db90eb5454508440da7be8c/src/sys/windows/tcp.rs#L36-L46

So the implementation should match every `mem::forget()` to a `overlapped2arc!()` to catch the struct and keep good count of the memory.

I added a trace debug for `FromRawArc`'s `clone` and `drop`, and to `UdpSocket`'s `drop` and indeed there seem to be a situation where refcount > 1 when UdpSocket is finally dropped:

```
[2019-02-25T14:51:07Z TRACE mio::poll] registering with poller
[2019-02-25T14:51:07Z TRACE mio::sys::windows::selector] register Token(0) Readable | Writable
[2019-02-25T14:51:07Z TRACE mio::sys::windows::selector] set readiness to (empty)
[2019-02-25T14:51:07Z TRACE mio::sys::windows::udp] scheduling a read
[2019-02-25T14:51:07Z TRACE mio::sys::windows::from_raw_arc] FromRawArc clone to cnt 2
[2019-02-25T14:51:07Z TRACE mio::sys::windows::selector] set readiness to Writable
[2019-02-25T14:51:07Z TRACE mio::sys::windows::selector] set readiness to (empty)
[2019-02-25T14:51:07Z TRACE mio::sys::windows::udp] scheduling a send
[2019-02-25T14:51:07Z TRACE mio::sys::windows::from_raw_arc] FromRawArc clone to cnt 3
[2019-02-25T14:51:08Z TRACE mio::sys::windows::udp] UdpSocket Drop
[2019-02-25T14:51:08Z TRACE mio::sys::windows::udp] cancelling active UDP read
[2019-02-25T14:51:08Z TRACE mio::sys::windows::from_raw_arc] FromRawArc drop to cnt 2
```

Since the refcount is still over1, the inners of `UdpSocket` are not dropped.

Also it looks like UDP's `send_done()` nor `recv_done()` are never called (there should be "finished a send|recv" debug trace). These are the two functions with `overlapped2arc!()` calls in UDP.

**Comment by oherrala** [comment]
This might be related to #776.

**Comment by Thomasdezeeuw** [comment]
@PerfectLaugh or @oherrala can either of you conform the memory leak doesn't exists on current master? I don't this is does any more as #776 is resolved.

**Comment by oherrala** [comment]
I assume it's this PR: https://github.com/tokio-rs/mio/pull/1034

Adding it here just for cross referencing.

**Comment by oherrala** [comment]
Thanks @PerfectLaugh for working on this!

I quickly tested the current master (e134e798d8f43f4e647dc14eb21464d7bb65132c) in Windows 10 using Dr. Memory 2.2.0 and my previous test code (adjusted for new mio) and Dr. Memory seems to be happy. No more leak like what was found above.

My program:

```rust
use std::io;
use std::net::{Ipv4Addr, SocketAddr};
use std::thread::sleep;
use std::time::Duration;

use env_logger;
use mio::net::UdpSocket;
use mio::{Interests, Poll, Token};

fn main() -> io::Result<()> {
    env_logger::init();

    let target = SocketAddr::new(Ipv4Addr::new(192, 168, 1, 1).into(), 53);
    let buf = &[0x00; 1400];
    let socket = UdpSocket::bind(SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0))?;

    let poll = Poll::new()?;
    poll.registry()
        .register(&socket, Token(0), Interests::READABLE)
        .unwrap();

    socket.send_to(buf, target)?;
    sleep(Duration::from_secs(1));

    Ok(())
}
```

**Comment by oherrala** [comment]
@Thomasdezeeuw @PerfectLaugh is there any change to get this backported into mio 0.6?

Trust-DNS on Windows is affected on this issue as well as probably other users of Tokio in Windows. Without fix in mio 0.6 I'm afraid it might take a while to get the ecosystem updated and fix deployed.

Issues hopefully solved by the fix:
 * https://github.com/bluejekyll/trust-dns/issues/692
 * https://github.com/seanmonstar/reqwest/issues/284
 * https://github.com/exonum/exonum/issues/628
 * https://github.com/stepancheg/grpc-rust/issues/110
 * https://github.com/shadowsocks/shadowsocks-rust/issues/146

**Comment by carllerche** [comment]
It could be possible to backport, but I would like the code to mature some on 0.7 first. There could be other bugs introduced in the change given that it is a rewrite. 

The more testing we get on 0.7, the sooner we can backport :-)

**Comment by Thomasdezeeuw** [comment]
Closing this for v0.7 and moving it to a new *maybe* backport to v0.6 milestone.

**Comment by PerfectLaugh** [comment]
> @Thomasdezeeuw @PerfectLaugh is there any change to get this backported into mio 0.6?
> 
> Trust-DNS on Windows is affected on this issue as well as probably other users of Tokio in Windows. Without fix in mio 0.6 I'm afraid it might take a while to get the ecosystem updated and fix deployed.
> 
> Issues hopefully solved by the fix:
> 
> * [bluejekyll/trust-dns#692](https://github.com/bluejekyll/trust-dns/issues/692)
> * [seanmonstar/reqwest#284](https://github.com/seanmonstar/reqwest/issues/284)
> * [exonum/exonum#628](https://github.com/exonum/exonum/issues/628)
> * [stepancheg/grpc-rust#110](https://github.com/stepancheg/grpc-rust/issues/110)
> * [shadowsocks/shadowsocks-rust#146](https://github.com/shadowsocks/shadowsocks-rust/issues/146)

I have made a backport. But due to 0.7 only supports edge-triggered mode. Level-triggered in the backport will cause error

**Comment by oherrala** [comment]
Congratulation on 0.7.0 release!

I tested the release with following program:

```toml
[package]
name = "miotest"
version = "0.1.0"
authors = ["Ossi Herrala <oherrala@iki.fi>"]
edition = "2018"

[dependencies]
env_logger = "0.7"
mio = { version = "0.7", features = [ "os-poll", "udp" ] }
```

```rust
use std::io;
use std::net::{Ipv4Addr, SocketAddr};
use std::thread::sleep;
use std::time::Duration;

use env_logger;
use mio::{Interest, Poll, Token};
use mio::net::UdpSocket;

fn main() -> io::Result<()> {
    env_logger::init();

    let target = SocketAddr::new(Ipv4Addr::new(192, 168, 1, 1).into(), 53);
    let buf = &[0x00; 1400];
    let mut socket = UdpSocket::bind(SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0))?;

    let poll = Poll::new()?;
    poll.registry()
        .register(&mut socket, Token(0), Interest::READABLE)
        .unwrap();

    socket.send_to(buf, target)?;
    sleep(Duration::from_secs(1));

    Ok(())
}
```

And Dr. Memory 2.3.0 reported no issues.

I'll close this issue and hope that upstreams pick up 0.7 instead of requiring a backport fix to 0.6.
